plugin-server/src/utils/lazy-loader.ts

```diff
@@ -0,0 +1,150 @@
+ * - Parallel loading defense - multiple calls for the same value in parallel only loads once
+ */
+
+export type LazyLoaderOptions<T> = {
+    name: string
+    /** Function to load the values */
+    loader: (key: string[]) => Promise<Record<string, T | undefined>>
+    /** How long to cache the value */
+    refreshAge?: number
+    /** How long to cache null values */
+    refreshNullAge?: number
+    /** How much jitter to add to the refresh time */
+    refreshJitterMs?: number
+    /** Whether to throw an error if the loader function throws an error */
+    throwOnLoadError?: boolean
+}
+
+export class LazyLoader<T> {
+    private cache: Record<string, T | null | undefined>
+    private lastUsed: Record<string, number | undefined>
+    private cacheUntil: Record<string, number | undefined>
+
+    constructor(private readonly options: LazyLoaderOptions<T>) {
+        this.cache = {}
+        this.lastUsed = {}
+        this.cacheUntil = {}
+    }
+
+    /**
+     * Ensure that a range of values are preloaded and cached.
+     *
+     * If already cached, the lastUsed value is updated to now
+     * If not cached, the value is loaded as part of the batch and added to the cache.
+     * If the value is older than the refreshAge, it is loaded from the database.
+     */
+    public async load(keys: string[]): Promise<Record<string, T | null>> {
+        const results: Record<string, T | null> = {}
+
+        const {
+            loader,
+            refreshAge = REFRESH_AGE,
+            refreshNullAge = REFRESH_AGE,
+            refreshJitterMs = REFRESH_JITTER_MS,
+            throwOnLoadError = true,
+        } = this.options
+        const keysToLoad = new Set<string>()
+
+        // First, check if all keys are already cached and update the lastUsed time
+        for (const key of keys) {
+            const cached = this.cache[key]
+
+            if (cached !== undefined) {
+                results[key] = cached
+                // Always update the lastUsed time
+                this.lastUsed[key] = Date.now()
+
+                const cacheUntil = this.cacheUntil[key] ?? 0
+
+                if (Date.now() > cacheUntil) {
+                    keysToLoad.add(key)
+                    lazyLoaderCacheHits.labels({ name: this.options.name, hit: 'miss' }).inc()
+                    continue
+                }
+            } else {
+                keysToLoad.add(key)
+                lazyLoaderCacheHits.labels({ name: this.options.name, hit: 'miss' }).inc()
+                continue
+            }
+
+            lazyLoaderCacheHits.labels({ name: this.options.name, hit: 'hit' }).inc()
+        }
+
+        if (keysToLoad.size === 0) {
+            lazyLoaderFullCacheHits.labels({ name: this.options.name, hit: 'hit' }).inc()
+            return results
+        }
+
+        lazyLoaderFullCacheHits.labels({ name: this.options.name, hit: 'miss' }).inc()
+
+        let loaded: Record<string, T | undefined>
```

Comment:
question: shouldn't this be `T | null`?

Reply:
I missed that accessing `Record<string, T>` doesn't return `T | undefined` for the values :S